From 9726175ac169c1ac94b15af2af7ea466feae6693 Mon Sep 17 00:00:00 2001
From: zesstra <zesstra@zesstra.de>
Date: Thu, 17 Sep 2009 16:35:11 +0200
Subject: [PATCH] Added efun system() to execute commands on the host system.

---
 src/efuns.c   |   26 ++++++++++++++++++++++++++
 src/func_spec |    2 +-
 2 files changed, 27 insertions(+), 1 deletions(-)

diff --git a/src/efuns.c b/src/efuns.c
index a718626..3b8e49f 100644
--- a/src/efuns.c
+++ b/src/efuns.c
@@ -9500,6 +9500,32 @@ v_strftime(svalue_t *sp, int num_arg)
     return sp;
 } /* f_strftime() */
 
+svalue_t *
+f_system(svalue_t *sp)
+
+/* EFUN system()
+ *
+ *     string system(string command)
+ *
+ * Execute <command> by calling system() on the host system.
+ */
+
+{
+    int res;
+    if (sp->type == T_NUMBER && sp->u.number == 0)
+    {
+        res = system(NULL);
+    }
+    else
+    {
+        res = system(get_txt(sp->u.str));
+    }
+    free_svalue(sp);
+    put_number(sp, res);
+
+    return sp;
+} /* f_system() */
+
 /***************************************************************************/
 
 /*-------------------------------------------------------------------------*/
diff --git a/src/func_spec b/src/func_spec
index 34a3ca0..6c7278c 100644
--- a/src/func_spec
+++ b/src/func_spec
@@ -585,7 +585,7 @@ void    set_driver_hook(int, closure|int|string|mapping|string*);
 void    shutdown(int default: F_CONST0);
 int     trace(int);
 string  traceprefix(string|int);
-
+int     system(string|null);
 
         /* Inventories (TODO: should be optional) */
 
-- 
1.7.4.1

